arm64 core, the regs command get the wrong SP after nil pointer dereference

1. What version of Delve are you using (`dlv version`)? 
```
Delve Debugger
Version: 1.21.2
Build: $Id: 98f8ab2662d926245917ade2f2bb38277315c7fc $
```

2. What version of Go are you using? (`go version`)?
`go version go1.20.10 linux/arm64`

3. What operating system and processor architecture are you using?
```
cat /etc/redhat-release
CentOS Linux release 7.9.2009 (AltArch)
```
4. What did you do?
```
package main

func main() {
	map_value := make(map[string]*int)

	var value_1 int = 1
	var value_2 int = 2
	map_value["a"] = &value_1
	map_value["b"] = &value_2

	mod_map(map_value)
}

func mod_map(map_value map[string]*int) {
	p_value := map_value["c"]
	*p_value = 3
}
```

```
step 1: Compile the demo.go with debug information (go build -gcflags=all="-N -l" demo.go)
step 2: export GOTRACEBACK=crash, then run the executable file. Now wen can get the coredump file and the information of the program.
step 3: Compare the information of the program and dlv coredump information.
```

5. What did you expect to see?
When `dlv core demo core_demo_xx`，we expect to get the right SP for the coredump function.
**SP=0x4000192dc0 is printed for the function mod_map which contains nil pointer dereference.**
The following is the inforamtion of the program. The inforamtion contain the SP for each frame stack.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x68bc0]

goroutine 1 [running]:
panic({0x760c0, 0xd1300})
        /home/wsl/go/go1.20.10/src/runtime/panic.go:987 +0x354 fp=0x4000192d60 sp=0x4000192ca0 pc=0x3ee44
runtime.panicmem()
        /home/wsl/go/go1.20.10/src/runtime/panic.go:260 +0x48 fp=0x4000192d80 sp=0x4000192d60 pc=0x3da78
runtime.sigpanic()
        /home/wsl/go/go1.20.10/src/runtime/signal_unix.go:841 +0x1e0 fp=0x4000192db0 sp=0x4000192d80 pc=0x533f0
main.mod_map(0x4000160668)
        /home/wsl/demo2/demo_2.go:16 +0x40 fp=0x4000192e00 sp=0x4000192dc0 pc=0x68bc0
main.main()
        /home/wsl/demo2/demo_2.go:11 +0x158 fp=0x4000192f70 sp=0x4000192e00 pc=0x68b68
runtime.main()
        /home/wsl/go/go1.20.10/src/runtime/proc.go:250 +0x1d4 fp=0x4000192fd0 sp=0x4000192f70 pc=0x41584
runtime.goexit()
        /home/wsl/go/go1.20.10/src/runtime/asm_arm64.s:1172 +0x4 fp=0x4000192fd0 sp=0x4000192fd0 pc=0x66584
```

6. What did you see instead?
Dlv get the wrong SP information for the function mod_map which contains nil pointer dereference.
**Dlv get he wrong SP = 0x4000192db0. The right SP is 0x4000192dc0.**
**We guess the wrong SP is the root cuase of the problem that the print commands for input parameter and local variable  is wrong.**

```
Type 'help' for list of commands.
(dlv) bt
 0  0x00000000000673d8 in runtime.raise
    at /home/wsl/go/go1.20.10/src/runtime/sys_linux_arm64.s:158
 1  0x0000000000053474 in runtime.dieFromSignal
    at /home/wsl/go/go1.20.10/src/runtime/signal_unix.go:883
 2  0x000000000005395c in runtime.sigfwdgo
    at /home/wsl/go/go1.20.10/src/runtime/signal_unix.go:1096
 3  0x00000000000526ac in runtime.sigtrampgo
    at /home/wsl/go/go1.20.10/src/runtime/signal_unix.go:432
 4  0x00000000000673d8 in runtime.raise
    at /home/wsl/go/go1.20.10/src/runtime/sys_linux_arm64.s:157
 5  0x0000000000053474 in runtime.dieFromSignal
    at /home/wsl/go/go1.20.10/src/runtime/signal_unix.go:883
 6  0x00000000000535e4 in runtime.crash
    at /home/wsl/go/go1.20.10/src/runtime/signal_unix.go:975
 7  0x000000000003f5a0 in runtime.fatalpanic
    at /home/wsl/go/go1.20.10/src/runtime/panic.go:1172
 8  0x000000000003ee44 in runtime.gopanic
    at /home/wsl/go/go1.20.10/src/runtime/panic.go:987
 9  0x000000000003da78 in runtime.panicmem
    at /home/wsl/go/go1.20.10/src/runtime/panic.go:260
10  0x00000000000533f0 in runtime.sigpanic
    at /home/wsl/go/go1.20.10/src/runtime/signal_unix.go:841
11  0x0000000000068bc0 in main.mod_map
    at ./demo_2.go:15
12  0x0000000000068bc0 in main.mod_map
    at ./demo_2.go:15
13  0x0000000000068bb8 in main.mod_map
    at ./demo_2.go:15
14  0x0000000000000000 in ???
    at :0
    error: NULL address
(truncated)
(dlv) frame 11
> runtime.raise() /home/wsl/go/go1.20.10/src/runtime/sys_linux_arm64.s:158 (PC: 0x673d8)
Warning: debugging optimized function
Frame 11: ./demo_2.go:15 (PC: 68bc0)
    10:
    11:         mod_map(map_value)
    12: }
    13:
    14: func mod_map(map_value map[string]*int) {
=>  15:         p_value := map_value["c"]
    16:         *p_value = 3
    17: }
(dlv) regs
 PC = 0x0000000000068bc0
 SP = 0x0000004000192db0
X29 = 0x00000040001605f8
X30 = 0x0000000000068bc0
```

--- a/_fixtures/issue3591.go
+++ b/_fixtures/issue3591.go
@@ -1,0 +1,18 @@
+// Fixture for issue #3591: nil pointer crash to test SP calculation in core dumps.
+package main
+
+func main() {
+	map_value := make(map[string]*int)
+
+	var value_1 int = 1
+	var value_2 int = 2
+	map_value["a"] = &value_1
+	map_value["b"] = &value_2
+
+	mod_map(map_value)
+}
+
+func mod_map(map_value map[string]*int) {
+	p_value := map_value["c"]
+	*p_value = 3
+}

--- a/pkg/proc/core/core_test.go
+++ b/pkg/proc/core/core_test.go
@@ -10,7 +10,9 @@ import (
 	"path"
 	"path/filepath"
 	"reflect"
+	"regexp"
 	"runtime"
+	"strconv"
 	"strings"
 	"testing"
 
@@ -543,5 +545,115 @@ func mustSupportCore(t *testing.T) {
 
 	if os.Getenv("CI") == "true" && buildMode == "pie" {
 		t.Skip("disabled on linux, Github Actions, with PIE buildmode")
+	}
+}
+
+// parseIssue3591RuntimeModMapSP parses runtime's SP for main.mod_map from panic output.
+func parseIssue3591RuntimeModMapSP(stderr []byte) (uint64, error) {
+	i := bytes.Index(stderr, []byte("main.mod_map("))
+	if i < 0 {
+		return 0, fmt.Errorf("main.mod_map not found in panic output:\n%s", stderr)
+	}
+	window := stderr[i:]
+	if len(window) > 2048 {
+		window = window[:2048]
+	}
+	m := regexp.MustCompile(`sp=(0x[0-9a-f]+)`).FindSubmatch(window)
+	if len(m) < 2 {
+		return 0, fmt.Errorf("sp= not found after main.mod_map in panic output:\n%s", window)
+	}
+	return strconv.ParseUint(string(m[1])[2:], 16, 64)
+}
+
+// TestIssue3591_CoreModMapFrameSP verifies Delve reports the same SP as the
+// Go runtime for sigpanic frames in linux/arm64 core dumps.
+func TestIssue3591_CoreModMapFrameSP(t *testing.T) {
+	if runtime.GOOS != "linux" || runtime.GOARCH != "arm64" {
+		t.Skip("issue #3591 reproduces on linux/arm64 ELF cores only")
+	}
+	mustSupportCore(t)
+
+	tempDir := t.TempDir()
+	var buildFlags test.BuildFlags
+	if buildMode == "pie" {
+		buildFlags = test.BuildModePIE
+	}
+	fix := test.BuildFixture(t, "issue3591", buildFlags)
+
+	panicPath := filepath.Join(tempDir, "panic.txt")
+	bashCmd := fmt.Sprintf("cd %q && ulimit -c unlimited && GOTRACEBACK=crash %q 2>%q",
+		tempDir, fix.Path, panicPath)
+	if out, err := exec.Command("bash", "-c", bashCmd).CombinedOutput(); len(out) > 0 {
+		t.Logf("crash run output: %s", out)
+		_ = err // process exits non-zero on crash
+	}
+
+	stderr, err := os.ReadFile(panicPath)
+	if err != nil || len(stderr) == 0 {
+		t.Fatalf("reading panic output: err=%v len=%d", err, len(stderr))
+	}
+	wantSP, err := parseIssue3591RuntimeModMapSP(stderr)
+	if err != nil {
+		t.Fatalf("%v\n--- panic.txt ---\n%s", err, stderr)
+	}
+
+	cores, err := filepath.Glob(path.Join(tempDir, "core*"))
+	switch {
+	case err != nil || len(cores) > 1:
+		t.Fatalf("core glob: err=%v cores=%v", err, cores)
+	case len(cores) == 0:
+		// Write outside t.TempDir(): global fixture cleanup runs after the temp
+		// directory is removed (see withCoreFile in this package).
+		cores = []string{fix.Path + ".issue3591.coredump"}
+		err := exec.Command("coredumpctl", "--output="+cores[0], "dump", fix.Path).Run()
+		if err != nil {
+			t.Skipf("no core in cwd and coredumpctl failed: %v", err)
+		}
+		test.AddPathToRemove(cores[0])
+	}
+	corePath := cores[0]
+
+	grp, err := OpenCore(corePath, fix.Path, []string{})
+	if err != nil {
+		t.Fatalf("OpenCore(%q): %v", corePath, err)
+	}
+	t.Cleanup(func() { grp.Detach(true) })
+	p := grp.Selected
+
+	gs, _, err := proc.GoroutinesInfo(p, 0, 0)
+	if err != nil || len(gs) == 0 {
+		t.Fatalf("GoroutinesInfo: err=%v n=%d", err, len(gs))
+	}
+	var g *proc.G
+	for _, gi := range gs {
+		if gi.ID == 1 {
+			g = gi
+			break
+		}
+	}
+	if g == nil {
+		t.Fatalf("goroutine 1 not found among %d goroutines", len(gs))
+	}
+
+	stack, err := proc.GoroutineStacktrace(p, g, 80, 0)
+	if err != nil {
+		t.Fatalf("GoroutineStacktrace: %v", err)
+	}
+	var modMap *proc.Stackframe
+	for i := range stack {
+		fn := stack[i].Current.Fn
+		if fn != nil && fn.Name == "main.mod_map" {
+			modMap = &stack[i]
+			break
+		}
+	}
+	if modMap == nil {
+		t.Fatalf("main.mod_map not in stack (%d frames)", len(stack))
+	}
+
+	gotSP := uint64(modMap.Regs.SP())
+	if gotSP != wantSP {
+		t.Fatalf("issue #3591: main.mod_map frame SP mismatch: delve=%#x runtime=%#x (see panic.txt in test temp dir pattern)",
+			gotSP, wantSP)
 	}
 }

--- a/pkg/proc/core/core_test.go
+++ b/pkg/proc/core/core_test.go
@@ -200,7 +200,7 @@ func TestSplicedReader(t *testing.T) {
 	}
 }
 
-func withCoreFile(t *testing.T, name, args string) *proc.TargetGroup {
+func withCoreFile(t *testing.T, name, args string) (*proc.TargetGroup, []byte) {
 	// This is all very fragile and won't work on hosts with non-default core patterns.
 	// Might be better to check in the binary and core?
 	tempDir := t.TempDir()
@@ -210,7 +210,7 @@ func withCoreFile(t *testing.T, name, args string) *proc.TargetGroup {
 	}
 	fix := test.BuildFixture(t, name, buildFlags)
 	bashCmd := fmt.Sprintf("cd %v && ulimit -c unlimited && GOTRACEBACK=crash %v %s", tempDir, fix.Path, args)
-	exec.Command("bash", "-c", bashCmd).Run()
+	output, _ := exec.Command("bash", "-c", bashCmd).CombinedOutput()
 	cores, err := filepath.Glob(path.Join(tempDir, "core*"))
 	switch {
 	case err != nil || len(cores) > 1:
@@ -220,7 +220,7 @@ func withCoreFile(t *testing.T, name, args string) *proc.TargetGroup {
 		err := exec.Command("coredumpctl", "--output="+cores[0], "dump", fix.Path).Run()
 		if err != nil {
 			t.Skipf("core file was not produced, could not run test, coredumpctl error: %v", err)
-			return nil
+			return nil, nil
 		}
 		test.AddPathToRemove(cores[0])
 	}
@@ -235,7 +235,7 @@ func withCoreFile(t *testing.T, name, args string) *proc.TargetGroup {
 		t.Errorf("read apport log: %q, %v", apport, err)
 		t.Fatalf("previous errors")
 	}
-	return p
+	return p, output
 }
 
 func logRegisters(t *testing.T, regs proc.Registers, arch *proc.Arch) {
@@ -256,7 +256,7 @@ func TestCore(t *testing.T) {
 
 	mustSupportCore(t)
 
-	grp := withCoreFile(t, "panic", "")
+	grp, _ := withCoreFile(t, "panic", "")
 	p := grp.Selected
 
 	recorded, _ := grp.Recorded()
@@ -333,7 +333,7 @@ func TestCoreFpRegisters(t *testing.T) {
 		t.Skip("not supported in go1.10 and later")
 	}
 
-	grp := withCoreFile(t, "fputest/", "panic")
+	grp, _ := withCoreFile(t, "fputest/", "panic")
 	p := grp.Selected
 
 	gs, _, err := proc.GoroutinesInfo(p, 0, 0)
@@ -414,7 +414,7 @@ func TestCoreWithEmptyString(t *testing.T) {
 	t.Parallel()
 	mustSupportCore(t)
 
-	grp := withCoreFile(t, "coreemptystring", "")
+	grp, _ := withCoreFile(t, "coreemptystring", "")
 	p := grp.Selected
 
 	gs, _, err := proc.GoroutinesInfo(p, 0, 0)
@@ -573,52 +573,13 @@ func TestIssue3591_CoreModMapFrameSP(t *testing.T) {
 	}
 	mustSupportCore(t)
 
-	tempDir := t.TempDir()
-	var buildFlags test.BuildFlags
-	if buildMode == "pie" {
-		buildFlags = test.BuildModePIE
-	}
-	fix := test.BuildFixture(t, "issue3591", buildFlags)
+	grp, stderr := withCoreFile(t, "issue3591", "")
+	p := grp.Selected
 
-	panicPath := filepath.Join(tempDir, "panic.txt")
-	bashCmd := fmt.Sprintf("cd %q && ulimit -c unlimited && GOTRACEBACK=crash %q 2>%q",
-		tempDir, fix.Path, panicPath)
-	if out, err := exec.Command("bash", "-c", bashCmd).CombinedOutput(); len(out) > 0 {
-		t.Logf("crash run output: %s", out)
-		_ = err // process exits non-zero on crash
-	}
-
-	stderr, err := os.ReadFile(panicPath)
-	if err != nil || len(stderr) == 0 {
-		t.Fatalf("reading panic output: err=%v len=%d", err, len(stderr))
-	}
 	wantSP, err := parseIssue3591RuntimeModMapSP(stderr)
 	if err != nil {
-		t.Fatalf("%v\n--- panic.txt ---\n%s", err, stderr)
+		t.Fatalf("%v\n--- crash output ---\n%s", err, stderr)
 	}
-
-	cores, err := filepath.Glob(path.Join(tempDir, "core*"))
-	switch {
-	case err != nil || len(cores) > 1:
-		t.Fatalf("core glob: err=%v cores=%v", err, cores)
-	case len(cores) == 0:
-		// Write outside t.TempDir(): global fixture cleanup runs after the temp
-		// directory is removed (see withCoreFile in this package).
-		cores = []string{fix.Path + ".issue3591.coredump"}
-		err := exec.Command("coredumpctl", "--output="+cores[0], "dump", fix.Path).Run()
-		if err != nil {
-			t.Skipf("no core in cwd and coredumpctl failed: %v", err)
-		}
-		test.AddPathToRemove(cores[0])
-	}
-	corePath := cores[0]
-
-	grp, err := OpenCore(corePath, fix.Path, []string{})
-	if err != nil {
-		t.Fatalf("OpenCore(%q): %v", corePath, err)
-	}
-	t.Cleanup(func() { grp.Detach(true) })
-	p := grp.Selected
 
 	gs, _, err := proc.GoroutinesInfo(p, 0, 0)
 	if err != nil || len(gs) == 0 {

--- a/pkg/proc/stack.go
+++ b/pkg/proc/stack.go
@@ -778,6 +778,14 @@ func (it *stackIterator) advanceRegsDWARF() (callFrameRegs op.DwarfRegisters, re
 		}
 	}
 
+	// Issue #3591: On LR architectures, runtime.sigpanic's DWARF CFA points to the
+	// signaled frame's SP, not the caller's SP. Add 2 words (saved FP + saved LR)
+	// to compute the correct caller SP.
+	if fn := it.bi.PCToFunc(it.pc); fn != nil && fn.Name == "runtime.sigpanic" && it.bi.Arch.usesLR {
+		callerSP := uint64(it.regs.CFA) + uint64(2*it.bi.Arch.PtrSize())
+		callFrameRegs.AddReg(callFrameRegs.SPRegNum, op.DwarfRegisterFromUint64(callerSP))
+	}
+
 	if it.bi.Arch.usesLR {
 		if ret == 0 && it.regs.Reg(it.regs.LRRegNum) != nil {
 			ret = it.regs.Reg(it.regs.LRRegNum).Uint64Val

